File: process/engines/itemrec/algorithms/scala/mahout/src/main/scala/io/prediction/algorithms/mahout/itemrec/svdplusplus/SVDPlusPlus.scala
package io.prediction.algorithms.mahout.itemrec.svdplusplus

import scala.collection.JavaConversions._

import io.prediction.algorithms.mahout.itemrec.MahoutJob

import org.apache.mahout.cf.taste.model.DataModel
import org.apache.mahout.cf.taste.recommender.Recommender
import org.apache.mahout.cf.taste.impl.recommender.svd.SVDRecommender
import org.apache.mahout.cf.taste.impl.recommender.svd.SVDPlusPlusFactorizer
import org.apache.mahout.cf.taste.impl.recommender.svd.Factorizer

class SVDPlusPlusJob extends MahoutJob {

  override def buildRecommender(dataModel: DataModel, args: Map[String, String]): Recommender = {

    val numFeatures: Int = getArg(args, "numFeatures").toInt
    val learningRate: Double = getArg(args, "learningRate").toDouble
    val preventOverfitting: Double = getArg(args, "preventOverfitting").toDouble
    val randomNoise: Double = getArg(args, "randomNoise").toDouble
    val numIterations: Int = getArg(args, "numIterations").toInt
    val learningRateDecay: Double = getArg(args, "learningRateDecay").toDouble

    val factorizer: Factorizer = new SVDPlusPlusFactorizer(dataModel, numFeatures, learningRate, preventOverfitting,
      randomNoise, numIterations, learningRateDecay)

    val recommender: Recommender = new SVDRecommender(dataModel, factorizer)

    recommender
  }
}

File: engines/src/main/java/engines/java/itemrec/algos/SVDPlusPlus.java
package io.prediction.engines.java.itemrec.algos;

import io.prediction.engines.java.itemrec.data.TrainingData;

import org.apache.mahout.cf.taste.model.DataModel;
import org.apache.mahout.cf.taste.recommender.Recommender;
import org.apache.mahout.cf.taste.impl.recommender.svd.SVDRecommender;
import org.apache.mahout.cf.taste.impl.recommender.svd.SVDPlusPlusFactorizer;
import org.apache.mahout.cf.taste.impl.recommender.svd.Factorizer;
import org.apache.mahout.cf.taste.common.TasteException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class SVDPlusPlus
  extends MahoutAlgorithm<SVDPlusPlusParams> {

  final static Logger logger = LoggerFactory.getLogger(SVDPlusPlus.class);

  @Override
  public Logger getLogger() {
    return logger;
  }

  @Override
  public Recommender buildRecommender(TrainingData cd) {
    // TODO: read from params
    int numFeatures = 3;
    double learningRate = 0.01;
    double preventOverfitting = 0.1;
    double randomNoise = 0.01;
    int numIterations = 3;
    double learningRateDecay = 1;
    boolean unseenOnly = false;

    Recommender recommender = null;
    // TODO: handle Exception
    try {
    Factorizer factorizer = new SVDPlusPlusFactorizer(
      cd.dataModel, numFeatures, learningRate, preventOverfitting,
      randomNoise, numIterations, learningRateDecay);

    // TODO: support other candidate item strategy
      recommender = new SVDRecommender(cd.dataModel, factorizer);
    } catch (TasteException e) {
      logger.error("Caught TasteException " + e.getMessage());
    }
    return recommender;
  }
}
